### DISCUSSION TITLE
Why have a got a "-alpha" semver from my release branch?

### DISCUSSION THREAD

**Discussion by tomqwpl** [GitTools/GitVersion#3604] [Q&A] [answered]
My first experiment had gone well. I had simulated a release cycle and got a 1.1.0 release, and then the develop branch started reporting 1.2.0 based version numbers, which was good. 

I then started simulating a second release cycle, and that's where things have gone awry. The git status is as follows:

![image](https://github.com/GitTools/GitVersion/assets/42344626/b9881c3d-c459-465a-b174-c89ea82f0bf2)

The tags have been added by the CI/CD process based on what gitversion reports. The master, develop and release/* branches get tagged.

The latest build of the "release/1.2" branch however is now reporting a version of 1.2.0-alpha.2. This is presumably because this is an explicit tag on this commit, despite that tag being added "to the develop branch" (added as a tag from a build of that branch). 

My GitVersion.yaml file is empty, so config is standard. The output from gitversion is:

```
INFO [06/19/23 14:37:35:17] Working directory: /builds/altairengineering/dma/slc/versiontest
INFO [06/19/23 14:37:35:19] Project root is: /builds/altairengineering/dma/slc/versiontest/
INFO [06/19/23 14:37:35:19] DotGit directory is: /builds/altairengineering/dma/slc/versiontest/.git
INFO [06/19/23 14:37:35:19] Branch from build environment: release/1.2
INFO [06/19/23 14:37:35:19] Begin: Normalizing git directory for branch 'release/1.2'
  INFO [06/19/23 14:37:35:24] One remote found (origin -> 'https://gitlab-ci-token:*******@gitlab.com/altairengineering/dma/slc/versiontest.git').
  INFO [06/19/23 14:37:35:24] Skipping fetching, if GitVersion does not calculate your version as expected you might need to allow fetching or use dynamic repositories
  INFO [06/19/23 14:37:35:25] Creating local branch release/1.2 pointing at ebece4e
  INFO [06/19/23 14:37:35:28] Creating local branch from remote tracking 'refs/remotes/origin/develop'.
  INFO [06/19/23 14:37:35:29] Creating local branch from remote tracking 'refs/remotes/origin/main'.
  INFO [06/19/23 14:37:35:29] Creating local branch from remote tracking 'refs/remotes/origin/release/1.1'.
  INFO [06/19/23 14:37:35:32] HEAD points at branch 'refs/heads/release/1.2'.
  INFO [06/19/23 14:37:35:32] End: Normalizing git directory for branch 'release/1.2' (Took: 128.25ms)
  INFO [06/19/23 14:37:35:34] Begin: Loading version variables from disk cache
    INFO [06/19/23 14:37:35:34] Cache file /builds/altairengineering/dma/slc/versiontest/.git/gitversion_cache/DF9F0C0EF0B073F7A2EEFF5A8DD7B31555242B47.yml not found.
  INFO [06/19/23 14:37:35:34] End: Loading version variables from disk cache (Took: 1.29ms)
  INFO [06/19/23 14:37:35:34] Using latest commit on specified branch
  INFO [06/19/23 14:37:35:44] Running against branch: release/1.2 (ebece4e Merge branch 'feature/anotherfeature' into 'develop')
  INFO [06/19/23 14:37:35:44] Current commit is tagged with version 1.2.0-alpha.2, version calculation is for metadata only.
  INFO [06/19/23 14:37:35:45] 0 commits found between ebece4e Merge branch 'feature/anotherfeature' into 'develop' and ebece4e Merge branch 'feature/anotherfeature' into 'develop'
  INFO [06/19/23 14:37:35:45] Begin: Calculating base versions
    INFO [06/19/23 14:37:35:47] Fallback base version: 0.1.0 with commit count source 0980c4efd1e52ffe855d3ebf8f[454](https://gitlab.com/altairengineering/dma/slc/versiontest/-/jobs/4499549158#L454)3a24681432b
    INFO [06/19/23 14:37:35:49] Git tag '1.2.0-alpha2': 1.2.0-alpha.2 with commit count source ebece4eca2ac2a62d9e78399751e502b45857699
    INFO [06/19/23 14:37:35:50] Begin: Finding branch source of 'release/1.2'
      INFO [06/19/23 14:37:35:51] Begin: Finding merge base between 'release/1.2' and 'develop'.
        INFO [06/19/23 14:37:35:52] Found merge base of ebece4e Merge branch 'feature/anotherfeature' into 'develop'
        INFO [06/19/23 14:37:35:52] Merge base of release/1.2' and 'develop is ebece4e Merge branch 'feature/anotherfeature' into 'develop'
      INFO [06/19/23 14:37:35:52] End: Finding merge base between 'release/1.2' and 'develop'. (Took: 12.23ms)
      INFO [06/19/23 14:37:35:52] Begin: Finding merge base between 'release/1.2' and 'main'.
        INFO [06/19/23 14:37:35:52] Found merge base of 6b694b5 Make change
        DEBUG [06/19/23 14:37:35:52] Second d7324c3 Merge branch 'develop' into 'main'
        DEBUG [06/19/23 14:37:35:52] New Merge base aae046e Make change
        INFO [06/19/23 14:37:35:52] Merge base was due to a forward merge, next merge base is aae046e Make change
        DEBUG [06/19/23 14:37:35:52] Second 0980c4e Initial commit
        DEBUG [06/19/23 14:37:35:52] New Merge base 0980c4e Initial commit
        INFO [06/19/23 14:37:35:52] Merge base was due to a forward merge, next merge base is 0980c4e Initial commit
        INFO [06/19/23 14:37:35:52] Merge base of release/1.2' and 'main is 0980c4e Initial commit
      INFO [06/19/23 14:37:35:52] End: Finding merge base between 'release/1.2' and 'main'. (Took: 2.21ms)
      INFO [06/19/23 14:37:35:52] Begin: Finding merge base between 'release/1.2' and 'release/1.1'.
        INFO [06/19/23 14:37:35:52] Found merge base of 6b694b5 Make change
        INFO [06/19/23 14:37:35:52] Merge base of release/1.2' and 'release/1.1 is 6b694b5 Make change
      INFO [06/19/23 14:37:35:52] End: Finding merge base between 'release/1.2' and 'release/1.1'. (Took: 0.47ms)
      INFO [06/19/23 14:37:35:53] Multiple source branches have been found, picking the first one (develop).
This may result in incorrect commit counting.
Options were:
develop, release/1.1, main
    INFO [06/19/23 14:37:35:53] End: Finding branch source of 'release/1.2' (Took: 33.02ms)
    INFO [06/19/23 14:37:35:53] Version in branch name: 1.2.0 with commit count source ebece4eca2ac2a62d9e78399751e502b[458](https://gitlab.com/altairengineering/dma/slc/versiontest/-/jobs/4499549158#L458)57699
    INFO [06/19/23 14:37:35:55] Found multiple base versions which will produce the same SemVer (1.2.0), taking oldest source for commit counting (Version in branch name)
    INFO [06/19/23 14:37:35:55] Base version used: Version in branch name: 1.2.0 with commit count source ebece4eca2ac2a62d9e78399751e502b45857699
  INFO [06/19/23 14:37:35:55] End: Calculating base versions (Took: 102.42ms)
  INFO [06/19/23 14:37:35:55] 0 commits found between ebece4e Merge branch 'feature/anotherfeature' into 'develop' and ebece4e Merge branch 'feature/anotherfeature' into 'develop'
  INFO [06/19/23 14:37:35:56] Begin: Getting version tags from branch 'refs/heads/release/1.2'.
  INFO [06/19/23 14:37:35:56] End: Getting version tags from branch 'refs/heads/release/1.2'. (Took: 5.40ms)
  INFO [06/19/23 14:37:35:57] Begin: Creating dictionary
  INFO [06/19/23 14:37:35:57] End: Creating dictionary (Took: 5.85ms)
  INFO [06/19/23 14:37:35:58] Begin: Storing version variables to cache file /builds/altairengineering/dma/slc/versiontest/.git/gitversion_cache/DF9F0C0EF0B073F7A2EEFF5A8DD7B31555242B47.yml
  INFO [06/19/23 14:37:35:63] End: Storing version variables to cache file /builds/altairengineering/dma/slc/versiontest/.git/gitversion_cache/DF9F0C0EF0B073F7A2EEFF5A8DD7B31555242B47.yml (Took: 55.84ms)
```

This seems to suggest that at least automatically applying tags to the "develop" build is a bad idea, or at least does an unexpected thing.

Any suggestions on how I modify my usage, configuration, or CI/CD workflow to produce a more expected output?

Thanks

**Comment by asbjornu** [comment]
What would you consider "a more expected output" in the given scenario?

**Reply by tomqwpl** [reply]
Naively, I've built code from the "release/1.2" branch. The gitversion configuration for that branch is that there should be a pre-release tag of "beta" on the version numbers. Instead, I've got an "-alpha" pre-release tag. 
Gitversion knows that the current branch is "release/1.2", yet I've got something with a different prerelease tag.

I'm just trying to understand how gitversion works and trying to construct some kind of human workflow and CI/CD pipeline for some kind of release process. Gitversion seems to be set up by default for "git flow" so I was trying to simulate that. At the moment it seems that I either need to stop automatically tagging builds on the develop branch or I need to add an artficial empty commit having created a release branch, otherwise builds from the release branch don't always get the right version number.

**Reply by asbjornu** [reply]
Automatically tagging non-stable releases seems superfluous, yes. You should preferably only need to tag stable releases.

**Reply by tomqwpl** [reply]
It takes notice of the "prerelease label" in the branch configuration in other circumstances:
![image](https://github.com/GitTools/GitVersion/assets/42344626/e846955d-55bc-4a88-a853-fcc3fe0a5935)

![image](https://github.com/GitTools/GitVersion/assets/42344626/2fbbc535-d1a7-43ce-a567-777bf35a0a1f)

So here, although the commit has a tag on it, the semver it's calculated has inherited the pre-release label from the branch config. This is different to what happened on my release branch.

Just for clarity. I created the feature branch feature/feature3, and the semver was taken from the build of that branch.

Note that here also, the "1.2" version confused me. Previously the develop branch automatically bumped to 1.2 when I had released 1.1. Here that doesn't seem to have happened.

**Reply by tomqwpl** [reply]
In terms of tagging builds on develop, the motivation was just to produce a "snapshot" or "nightly" build that can be downloaded and used by developers. This isn't uncommon. The natural way to do that is to create a tag, but I suppose it doesn't have to be a tag that is of the form "v1.2.3".

**Comment by tomqwpl** [comment]
Actually I notice that I got the same `-alpha` version on the release/1.1 branch as well.
This suggests that I would need to create a commit after creating the release/1.1 branch?, so that the tag is no longer on the same commit. That does indeed work.

**Reply by asbjornu** [reply]
Correct. When several branches point to the same commit, it's ambiguous to GitVersion which branch it should consider for version calculation.

**Reply by tomqwpl** [reply]
But git knows that the "current" branch is just one of those doesn't it?
![image](https://github.com/GitTools/GitVersion/assets/42344626/d98415fe-a7eb-4d58-9b81-d2d951e82c1a)
![image](https://github.com/GitTools/GitVersion/assets/42344626/95f3df46-975f-41fd-a79d-524d597e05ee)

I think that's probably the bit that's confusing to me here. It appears to be ignoring the fact that the current branch is the "release/1.3" branch and so not applying the gitversion branch configuration pre-release tag.

**Reply by HHobeck** [reply, marked as answer]
The feature branch in the gitflow workflow has a increment strategy set to inherit. Thus in your scenario it inherits from develop and release branch. I would recommend you to use the beta release: https://github.com/GitTools/GitVersion/releases/tag/6.0.0-beta.2

**Reply by tomqwpl** [reply]
#2340 appears to describe the situation I'm seeing, where a tag exists that doesn't match the configuration of the current branch.
I will check out the 6.0 version and see if that improves my scenario

**Reply by HHobeck** [reply]
It's a general problem. From the repository point of view the state is the same until you are creating an additional commit on develop or release branch:

```csharp
[TestCase("develop")]
[TestCase("release/1.1.0")]
public void __Just_A_Test__(string branch)
{
    var configuration = GitFlowConfigurationBuilder.New.Build();

    using var fixture = new EmptyRepositoryFixture("main");

    fixture.MakeATaggedCommit("1.0.0");

    fixture.BranchTo("develop");
    fixture.MakeACommit("1");

    fixture.BranchTo("release/1.1.0");

    // ✅ succeeds as expected
    fixture.AssertFullSemver("1.1.0-beta.1+0", configuration);

    fixture.Checkout("develop");

    // ✅ succeeds as expected
    fixture.AssertFullSemver("1.2.0-alpha.0", configuration);

    fixture.Checkout(branch);
    fixture.BranchTo("feature/foo");
    fixture.MakeACommit("2");

    // ✅ succeeds as expected
    fixture.AssertFullSemver("1.2.0-foo.1+1", configuration);
}
```

The scenario says if you create a feature branch from a develop or a release branch it makes no different when the release and the develop branch points to the same commit. It yields to the highest version which might be not expected.

**Reply by tomqwpl** [reply]
I'm not really sure I understand that test case. The only tag involved there is the initial 1.0.0 tag. So the versions it comes up with seem reasonable, but not obviously the scenario I'm thinking of.
Right now, the 6.0.0-beta-2 is failing in my CI/CD pipeline, it's apparently returning a non zero return code which I need to try and track down:
![image](https://github.com/GitTools/GitVersion/assets/42344626/70143702-e5ef-4839-a058-23cc248302d5)

![image](https://github.com/GitTools/GitVersion/assets/42344626/c2a0f0f8-8989-4a34-8fc4-0c8b2edf464e)

**Reply by HHobeck** [reply]
There are breaking changes in yaml configuration?

**Reply by HHobeck** [reply]
The scenario says if you create a feature branch from a develop or a release branch it makes no different when the release and the develop branch points to the same commit. It yields to the highest version which might be not expected.

**Reply by tomqwpl** [reply]
The breaking change was the removal of LegacySemVer. The example I copied extracted that rather than SemVer. Not sure why, not actually sure I know what the differences were. 
Anyway the behaviour with that change seems more expected to me, only take notice of a tag if it has the same "pre-release" part as the branch config says should be used for the current branch. I feel more confident in the output now. There's one oddity I still need to investigate, but I need to run through my test again with this new version.

**Reply by tomqwpl** [reply]
OK, that change in #2340 doesn't seem to have done what I have expected.
What I interpretted it was going to do (because this is behaviour I wanted in this particular circumstance) was that it would completely ignore the tag. So I have created a branch called "release/1.1" and the commit I created it from had a tag "0.1.0-alpha.3". The version number that gitversion has chosen is something based on 0.1.0 from the tag), rather than something based on 1.1 (from the release name). It has been given the right prerelease tag (-beta), and it's possible that was the change made.
I think what I was expecting (hoping) was that the entire presence of the tag would be ignored, as it didn't have a pre-release part that matches the current branch config.

**Reply by tomqwpl** [reply]
Ah. It matters that the branch is called "release/1.2" rather than "release/1.2.3".  This test fails:
```
        var configuration = GitFlowConfigurationBuilder.New
            .WithBranch("develop", builder => builder.WithLabel("alpha"))
            .WithBranch("release", builder => builder.WithLabel("beta"))
            .Build();

        using var fixture = new EmptyRepositoryFixture();

        fixture.MakeACommit();
        fixture.BranchTo("develop");
        fixture.MakeACommit();
        fixture.AssertFullSemver("0.1.0-alpha.2", configuration);
        fixture.MakeATaggedCommit("0.1.0-alpha.2");
        fixture.BranchTo("release/1.0");

        // ✅ succeeds as expected
        fixture.AssertFullSemver("1.0.0-beta.1+0", configuration);
```
Change the release name to "release/1.0.0" and the test works.

**Reply by asbjornu** [reply]
Correct. The default regex for version numbers in branch names is restrictive to not accidentally interpret any number in the branch name as a version number.

**Comment by HHobeck** [comment]
The title of this discussion is: `Why have a got a "-alpha" semver from my release branch?`

Could you please focus on this question and mark it as answered if it has been answered.

**Reply by tomqwpl** [reply]
But all of this is to do with that. Fundamentally it is solved by that 2340 fix

**Comment by HHobeck** [comment]
> What I interpretted it was going to do (because this is behaviour I wanted in this particular circumstance) was that it would completely ignore the tag. [..]
> I think what I was expecting (hoping) was that the entire presence of the tag would be ignored, as it didn't have a pre-release part that matches the current branch config.

May I ask you: Why do you expect the semantic version of a tag to be ignored just because the pre-release label is different on the branch? See [[Bug] Consider pre-release tags only when they matching with the label name on branch #3438](https://github.com/GitTools/GitVersion/issues/3438).

```csharp
[Test]
public void Just_A_Test()
{
    var configuration = GitFlowConfigurationBuilder.New.Build();

    using var fixture = new EmptyRepositoryFixture("main");

    fixture.MakeACommit();

    // ✅ succeeds as expected
    fixture.AssertFullSemver($"0.0.1-1", configuration);

    fixture.BranchTo("develop");
    fixture.MakeACommit();

    // ✅ succeeds as expected
    fixture.AssertFullSemver($"0.1.0-alpha.2", configuration);

    fixture.ApplyTag($"1.0.0-alpha.2");

    // ✅ succeeds as expected
    fixture.AssertFullSemver($"1.0.0-alpha.2", configuration);

    fixture.BranchTo("release/next");
    fixture.MakeACommit();

    // ✅ succeeds as expected
    fixture.AssertFullSemver($"1.0.0-beta.1+3", configuration);
}
```

**Comment by HHobeck** [comment]
> May I ask you: Why do you expect the semantic version of a tag to be ignored just because the pre-release label is different on the branch?

@tomqwpl

**Reply by tomqwpl** [reply]
Isn't that what that change implements though? It appears to at least. Indeed the test I wrote shows that behaviour, assuming I've written my test correctly:
```
 fixture.MakeACommit();
        fixture.BranchTo("develop");
        fixture.MakeACommit();
        fixture.AssertFullSemver("0.1.0-alpha.2", configuration);
        fixture.MakeATaggedCommit("0.1.0-alpha.2");
        fixture.BranchTo("release/1.0.0");
        fixture.AssertFullSemver("1.0.0-beta.1+0", configuration);
```

If you weren't using gitversion you might implement this part of the logic in your Jenkinsfile for example. You might write logic in your Jenkinsfile like 

```
if branchname ~= release/* then prereleasetag="-beta"
else branchname == develop then prereleasetag="-alpha"
else branchname == master then prereleasetag=""
```

I appreciate that it is possible to check out a tag and build the code from that, and it's then potentially ambiguous what the branch is, and so potentially make reproducing a build from the same commit ambiguous. I don't think I've ever rebuilt from a tag ever. The 99.9999% case is always "build the tip of branch xyz", and then tag that afterwards as "I built this and it generated a build with version blah".

**Reply by HHobeck** [reply]
Actually I'm not sure if I can follow your thoughts. Of course you can specify the pre-release label per branch. And of course the pre-release label of an tag will be ignored (and the pre-release number will be not reset) if the label on the branch differs. But the pre-released version number (which comes from a semantic version from a tag with a foreign pre-relesae label) will be cosidered. Are you agree? That means the expectation that the tag will be completely ignored is not correct.

**Expected:**
```csharp
[Test]
public void Just_A_Test()
{
    var configuration = GitFlowConfigurationBuilder.New.Build();

    using var fixture = new EmptyRepositoryFixture("main");

    fixture.MakeACommit();

    // ✅ succeeds as expected
    fixture.AssertFullSemver($"0.0.1-1", configuration);

    fixture.BranchTo("develop");
    fixture.MakeACommit();

    // ✅ succeeds as expected
    fixture.AssertFullSemver($"0.1.0-alpha.2", configuration);

    fixture.ApplyTag($"1.0.0-alpha.2");

    // ✅ succeeds as expected
    fixture.AssertFullSemver($"1.0.0-alpha.2", configuration);

    fixture.BranchTo("release/next");
    fixture.MakeACommit();

    // ✅ succeeds as expected
    fixture.AssertFullSemver($"1.0.0-beta.1+3", configuration);
}
```

**Not Expected:**
```csharp
[Test]
public void Just_A_Test()
{
    var configuration = GitFlowConfigurationBuilder.New.Build();

    using var fixture = new EmptyRepositoryFixture("main");

    fixture.MakeACommit();

    // ✅ succeeds as expected
    fixture.AssertFullSemver($"0.0.1-1", configuration);

    fixture.BranchTo("develop");
    fixture.MakeACommit();

    // ✅ succeeds as expected
    fixture.AssertFullSemver($"0.1.0-alpha.2", configuration);

    fixture.ApplyTag($"1.0.0-alpha.2");

    // ✅ succeeds as expected
    fixture.AssertFullSemver($"1.0.0-alpha.2", configuration);

    fixture.BranchTo("release/next");
    fixture.MakeACommit();

    /// ❌ expected: 1.0.0-beta.1+3
    fixture.AssertFullSemver($"0.1.0-beta.1+3", configuration); // <<--- the tag 1.0.0-alpha.2 should be considered but not the label name
}
```

**Reply by tomqwpl** [reply]
I'm confused where this thread has gone.
I originally raised an query saying "in these circumstances I would expect this behaviour". Upshot of the initial conversation appeared to be "that is the new behaviour in version 6.0.0". I tried 6.0.0 and it appeared to give the behaviour I was looking for,
The question then appeared to be raised "why would you expect that behaviour"? 
Unless I'm missing something, I _think_ 6.0.0 does exactly what I'm expecting, so end of story.

What am I missing?

**Reply by HHobeck** [reply]
Okay no problem. Just want to ensure that we are on the same page and the behavior in 6.0.0 is like you expecting it.
